msg.go

```diff
@@ -360,94 +376,116 @@ func isRootLabel(s string, bs []byte, off, end int) bool {
+	if err != nil {
+		if errors.Is(err, errUnpackOverflow) {
+			// Keep existing behaviour of returning ErrBuf here.
+			return "", len(msg), ErrBuf
+		}
+		// Keep documented behaviour of returning len(msg) here.
+		return "", len(msg), err
+	}
+	return name, offset(s, msg), nil
+}
+
+func unpackDomainName(s *cryptobyte.String, msgBuf []byte) (string, error) {
+	name := make([]byte, 0, maxDomainNamePresentationLength)
 	budget := maxDomainNameWireOctets
-	ptr := 0 // number of pointers followed
-Loop:
+	var ptrs int // number of pointers followed
+
+	// If we never see a pointer, we need to ensure that we advance s to our
+	// final position.
+	cs := *s
+	defer func() {
+		if ptrs == 0 {
+			*s = cs
+		}
+	}()
+
 	for {
-		if off >= lenmsg {
-			return "", lenmsg, ErrBuf
+		var c byte
+		if !cs.ReadUint8(&c) {
+			return "", errUnpackOverflow
 		}
-		c := int(msg[off])
-		off++
 		switch c & 0xC0 {
-		case 0x00:
-			if c == 0x00 {
-				// end of name
-				break Loop
+		case 0x00: // literal string
+			var label []byte
+			if !cs.ReadBytes(&label, int(c)) {
+				return "", errUnpackOverflow
 			}
-			// literal string
-			if off+c > lenmsg {
-				return "", lenmsg, ErrBuf
+			// If we see a zero-length label (root label), this is the
+			// end of the name.
+			if len(label) == 0 {
+				if len(name) == 0 {
+					return ".", nil
+				}
+				return string(name), nil
 			}
-			budget -= c + 1 // +1 for the label separator
-			if budget <= 0 {
-				return "", lenmsg, ErrLongDomain
+			if budget -= len(label) + 1; budget <= 0 { // +1 for the label separator
+				return "", ErrLongDomain
 			}
-			for _, b := range msg[off : off+c] {
+			for _, b := range label {
 				if isDomainNameLabelSpecial(b) {
-					s = append(s, '\\', b)
+					name = append(name, '\\', b)
 				} else if b < ' ' || b > '~' {
-					s = append(s, escapeByte(b)...)
+					name = append(name, escapeByte(b)...)
 				} else {
-					s = append(s, b)
+					name = append(name, b)
 				}
 			}
-			s = append(s, '.')
-			off += c
-		case 0xC0:
-			// pointer to somewhere else in msg.
-			// remember location after first ptr,
-			// since that's how many bytes we consumed.
-			// also, don't follow too many pointers --
-			// maybe there's a loop.
-			if off >= lenmsg {
-				return "", lenmsg, ErrBuf
+			name = append(name, '.')
+		case 0xC0: // pointer
+			var c1 byte
+			if !cs.ReadUint8(&c1) {
+				return "", errUnpackOverflow
 			}
-			c1 := msg[off]
-			off++
-			if ptr == 0 {
-				off1 = off
+			// If this is the first pointer we've seen, we need to
+			// advance s to our current position.
+			if ptrs == 0 {
+				*s = cs
 			}
-			if ptr++; ptr > maxCompressionPointers {
-				return "", lenmsg, &Error{err: "too many compression pointers"}
+			// Don't follow too many pointers in case there is a loop.
+			//
+			// TODO(tmthrgd): This check is no longer required to ensure
+			// loop free behaviour. Should we remove it?
+			if ptrs++; ptrs > maxCompressionPointers {
```

Comment:
technically there could still be a lot of pointers backwards? `maxCompressionsPointers` now has a value of 126; that could still be reach, although not via a loop.

Think it can go, even with a mesg with a lot of pointers backwards, we should not introduce an arbitrary limit

Reply:
Yeah it's a weird one. Pointing to another pointer would never be generated by a sensible non-buggy implementation. It doesn't even save you a byte. There's annoyingly no RFC that prohibits it though, so it is very much a judgement call. The backward check prevents a loop so we can remove this, it just opens the door to being forced to do more work by an attacker. I can remove it in a follow up PR.

Reply:
pointer to self is then still checked?